// wikijs-33/dist/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.api = api;
exports.pagination = pagination;
exports.aggregatePagination = aggregatePagination;

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

var _querystring = require('querystring');

var _querystring2 = _interopRequireDefault(_querystring);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var fetchOptions = {
	method: 'GET',
	headers: {
		'User-Agent': 'WikiJs/0.1 (https://github.com/dijs/wiki; richard.vanderdys@gmail.com)'
	}
};

function api(apiOptions) {
	var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

	var qs = Object.assign({}, params, {
		format: 'json',
		action: 'query',
		redirects: ''
	});
	var url = apiOptions.apiUrl + '?' + _querystring2.default.stringify(qs);
	return (0, _isomorphicFetch2.default)(url, fetchOptions).then(function (res) {
		return res.json();
	});
}

function pagination(apiOptions, params, parseResults) {
	return api(apiOptions, params).then(function (res) {
		var resolution = {};
		resolution.results = parseResults(res);
		if (res['continue']) {
			var continueType = Object.keys(res['continue']).filter(function (key) {
				return key !== 'continue';
			})[0];
			var continueKey = res['continue'][continueType];
			params[continueType] = continueKey;
			resolution.next = function () {
				return pagination(apiOptions, params, parseResults);
			};
		}
		return resolution;
	});
}

function aggregatePagination(pagination) {
	var previousResults = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];

	return pagination.then(function (res) {
		var results = [].concat(_toConsumableArray(previousResults), _toConsumableArray(res.results));
		if (res.next) {
			return aggregatePagination(res.next(), results);
		} else {
			return results;
		}
	});
}
